File: static/js/push.js
'use strict';

// Web Push opt-in: the "Activer les notifications push" toggle inside
// #user-settings-modal (static/index.html), wired to GET
// /api/v1/push/vapid-public-key and POST/DELETE /api/v1/push/subscribe
// (internal/handlers/push.go) plus the browser's own Notification/
// PushManager APIs. Shares `state`, `apiRequest`, `showError`/`hideError`,
// `t`, `selectList` with app.js/list_view.js/settings.js — same
// classic-<script>-tags shared-scope pattern as every other frontend file.
//
// refreshPushToggleUI is called from settings.js's openUserSettingsModal
// (every time the modal opens, not just once) since permission/subscription
// state can change outside the app at any time — the user revoking
// notification access via the browser's own site settings, most notably —
// and the toggle has to reflect reality rather than whatever it last showed.

const pushEls = {
  toggle: document.getElementById('user-settings-push-toggle'),
  status: document.getElementById('user-settings-push-status'),
};

function isPushSupported() {
  return 'serviceWorker' in navigator && 'PushManager' in window && 'Notification' in window;
}

// Converts a base64url-encoded VAPID public key (as
// GET /api/v1/push/vapid-public-key returns it) into the raw Uint8Array
// PushManager.subscribe's applicationServerKey option expects — the
// standard conversion every Web Push integration needs, since the Push API
// itself has no base64 convenience of its own.
function urlBase64ToUint8Array(base64String) {
  const padding = '='.repeat((4 - (base64String.length % 4)) % 4);
  const base64 = (base64String + padding).replace(/-/g, '+').replace(/_/g, '/');
  const raw = atob(base64);
  const bytes = new Uint8Array(raw.length);
  for (let i = 0; i < raw.length; i++) {
    bytes[i] = raw.charCodeAt(i);
  }
  return bytes;
}

// Cached for the page's lifetime — the VAPID public key never changes
// without a server restart, and both refreshPushToggleUI and enablePush
// need it.
let vapidKeyPromise = null;

function fetchVapidPublicKey() {
  if (!vapidKeyPromise) {
    vapidKeyPromise = apiRequest('/push/vapid-public-key').catch((err) => {
      vapidKeyPromise = null; // let a later call retry after a transient failure
      throw err;
    });
  }
  return vapidKeyPromise;
}

async function getExistingPushSubscription() {
  if (!isPushSupported()) return null;
  const registration = await navigator.serviceWorker.ready;
  return registration.pushManager.getSubscription();
}

// Sets the toggle's checked/disabled state and the small status line below
// it to match reality: unsupported browser, push not configured on this
// instance, permission previously denied (which only the user can undo, via
// their own browser's site settings — this app can't re-prompt), or the
// ordinary on/off state of an actual browser subscription.
async function refreshPushToggleUI() {
  if (!pushEls.toggle) return;
  pushEls.status.hidden = true;
  pushEls.status.textContent = '';

  if (!isPushSupported()) {
    pushEls.toggle.checked = false;
    pushEls.toggle.disabled = true;
    pushEls.status.textContent = t('modals.userSettings.pushUnsupported');
    pushEls.status.hidden = false;
    return;
  }

  let vapid;
  try {
    vapid = await fetchVapidPublicKey();
  } catch {
    vapid = { enabled: false };
  }
  if (!vapid.enabled) {
    pushEls.toggle.checked = false;
    pushEls.toggle.disabled = true;
    pushEls.status.textContent = t('modals.userSettings.pushNotConfigured');
    pushEls.status.hidden = false;
    return;
  }

  if (Notification.permission === 'denied') {
    pushEls.toggle.checked = false;
    pushEls.toggle.disabled = true;
    pushEls.status.textContent = t('modals.userSettings.pushPermissionDenied');
    pushEls.status.hidden = false;
    return;
  }

  pushEls.toggle.disabled = false;
  const subscription = await getExistingPushSubscription();
  pushEls.toggle.checked = !!subscription;
}

// Requests notification permission (the browser's own native prompt —
// nothing here can skip or pre-empt it) and, once granted, subscribes via
// PushManager and registers the subscription with the backend.
async function enablePush() {
  const vapid = await fetchVapidPublicKey();
  if (!vapid.enabled) {
    throw new Error(t('modals.userSettings.pushNotConfigured'));
  }

  const permission = await Notification.requestPermission();
  if (permission !== 'granted') {
    throw new Error(t('modals.userSettings.pushPermissionDenied'));
  }

  const registration = await navigator.serviceWorker.ready;
  let subscription = await registration.pushManager.getSubscription();
  if (!subscription) {
    subscription = await registration.pushManager.subscribe({
      userVisibleOnly: true,
      applicationServerKey: urlBase64ToUint8Array(vapid.public_key),
    });
  }

  const json = subscription.toJSON();
  await apiRequest('/push/subscribe', {
    method: 'POST',
    body: JSON.stringify({ endpoint: json.endpoint, keys: json.keys }),
  });
}

// Unsubscribes both locally (the browser stops being able to receive a push
// for this subscription at all) and on the backend (so the now-dead
// endpoint isn't kept around and tried on the next notification).
async function disablePush() {
  const subscription = await getExistingPushSubscription();
  if (!subscription) return;
  const endpoint = subscription.endpoint;
  await subscription.unsubscribe();
  try {
    await apiRequest('/push/subscribe', { method: 'DELETE', body: JSON.stringify({ endpoint }) });
  } catch {
    // The subscription is already gone client-side regardless — a failure
    // reaching the server here just means a stale row lingers until the
    // push service itself eventually reports it gone (see
    // internal/handlers.sendToUsers' ErrSubscriptionGone cleanup), which
    // isn't worth surfacing as an error for a toggle the user just turned
    // off.
  }
}

if (pushEls.toggle) {
  pushEls.toggle.addEventListener('change', async () => {
    hideError();
    pushEls.status.hidden = true;
    const wantEnabled = pushEls.toggle.checked;
    pushEls.toggle.disabled = true;
    try {
      if (wantEnabled) {
        await enablePush();
      } else {
        await disablePush();
      }
    } catch (err) {
      pushEls.toggle.checked = !wantEnabled;
      if (!isNetworkError(err)) showError(err.message);
    } finally {
      pushEls.toggle.disabled = false;
    }
  });
}

File: internal/handlers/push.go
package handlers

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"strings"
	"sync"
	"time"

	"trakka/internal/db"
	"trakka/internal/models"
	"trakka/internal/validate"
	"trakka/internal/webpush"
)

// pushSendTimeout bounds one whole notifyListChange/RunRecurringDueScan
// fan-out to every recipient's subscriptions — generous for a handful of
// household members' devices, short enough that a slow or hanging push
// service can't leak a goroutine indefinitely. Each individual
// webpush.Send call is additionally bounded by its own package-level
// http.Client timeout regardless.
const pushSendTimeout = 30 * time.Second

// subscriptionCleanupTimeout bounds the best-effort delete of a subscription
// the push service reported as permanently gone. It deliberately derives from
// context.Background() rather than reusing sendToUsers' own ctx, which may
// already be at or near pushSendTimeout's deadline by the time an individual
// delivery goroutine gets here — but it still needs its own bound rather than
// running fully unbounded, so a stuck DB call here can't leak the goroutine.
const subscriptionCleanupTimeout = 5 * time.Second

// ---------------------------------------------------------------------------
// Subscription management
// ---------------------------------------------------------------------------

// handlePushVAPIDPublicKey hands the frontend the public half of this
// instance's VAPID identity, so PushManager.subscribe()'s
// applicationServerKey never has to be hard-coded into the frontend or
// duplicated from VAPID_PUBLIC_KEY by hand — see static/js/push.js.
// {"enabled": false} (never a 404) when push isn't configured at all, so
// the frontend's settings toggle can just branch on the field rather than
// on response status.
func (app *Application) handlePushVAPIDPublicKey(w http.ResponseWriter, r *http.Request) {
	if !app.Config.PushEnabled() {
		writeJSON(w, http.StatusOK, map[string]any{"enabled": false})
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"enabled": true, "public_key": app.Config.VAPIDPublicKey})
}

// handlePushSubscribe registers (or refreshes) a Web Push subscription for
// the calling user, in the shape a browser's PushSubscription.toJSON()
// produces: {"endpoint": "...", "keys": {"p256dh": "...", "auth": "..."}}.
func (app *Application) handlePushSubscribe(w http.ResponseWriter, r *http.Request) {
	if !app.Config.PushEnabled() {
		writeError(w, http.StatusServiceUnavailable, "push notifications are not configured on this instance")
		return
	}

	var in struct {
		Endpoint string `json:"endpoint"`
		Keys     struct {
			P256dh string `json:"p256dh"`
			Auth   string `json:"auth"`
		} `json:"keys"`
	}
	if !decodeJSON(w, r, &in) {
		return
	}

	// Every real push service endpoint is a plain https URL — internal/validate.URL
	// accepts http too (it exists for item.url, which has no such
	// restriction), so an explicit scheme check is layered on top here, the
	// same defense-in-depth internal/webpush.Send itself re-applies before
	// ever dialing this value.
	cleanEndpoint, err := validate.URL(in.Endpoint)
	if err != nil {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}
	if cleanEndpoint == "" {
		writeError(w, http.StatusBadRequest, "endpoint is required")
		return
	}
	if !strings.HasPrefix(strings.ToLower(cleanEndpoint), "https://") {
		writeError(w, http.StatusBadRequest, "endpoint must be an https:// URL")
		return
	}
	if in.Keys.P256dh == "" || in.Keys.Auth == "" {
		writeError(w, http.StatusBadRequest, "keys.p256dh and keys.auth are required")
		return
	}

	sub, err := app.DB.CreatePushSubscription(r.Context(), userFromContext(r).ID, cleanEndpoint, in.Keys.P256dh, in.Keys.Auth, r.UserAgent())
	if err != nil {
		app.serverError(w, r, err)
		return
	}
	writeJSON(w, http.StatusCreated, sub)
}

// handlePushUnsubscribe removes a subscription for the calling user —
// called both on an explicit "disable push" toggle and defensively before a
// fresh subscribe (see static/js/push.js), so a stale endpoint can never
// linger alongside a new one for the same device.
func (app *Application) handlePushUnsubscribe(w http.ResponseWriter, r *http.Request) {
	var in struct {
		Endpoint string `json:"endpoint"`
	}
	if !decodeJSON(w, r, &in) {
		return
	}
	if in.Endpoint == "" {
		writeError(w, http.StatusBadRequest, "endpoint is required")
		return
	}

	if err := app.DB.DeletePushSubscription(r.Context(), userFromContext(r).ID, in.Endpoint); errors.Is(err, db.ErrNotFound) {
		// Not finding it is not a failure from the caller's point of view —
		// the end state ("this endpoint is not subscribed") is exactly what
		// was asked for either way, so this is intentionally idempotent
		// rather than a 404.
		w.WriteHeader(http.StatusNoContent)
		return
	} else if err != nil {
		app.serverError(w, r, err)
		return
	}
	w.WriteHeader(http.StatusNoContent)
}

// handlePushTest sends one push notification to every subscription the
// calling user has registered, as an end-to-end diagnostic: unlike
// notifyListChange/RunRecurringDueScan, whose delivery is always a
// best-effort side effect of some other action, this endpoint's entire
// purpose is the delivery itself, so it runs synchronously (bounded by
// pushSendTimeout, same as every other fan-out in this file) rather than in
// a detached goroutine, and reports back how many subscriptions it attempted
// rather than silently succeeding either way. Scoped to the caller's own
// account only — there is no way to test-notify anyone else — so no
// house/list access check applies, only the ordinary RequireSession gate
// every /api/v1/... route already sits behind.
func (app *Application) handlePushTest(w http.ResponseWriter, r *http.Request) {
	if !app.Config.PushEnabled() {
		writeError(w, http.StatusServiceUnavailable, "push notifications are not configured on this instance")
		return
	}

	user := userFromContext(r)
	subs, err := app.DB.ListPushSubscriptionsForUsers(r.Context(), []int64{user.ID})
	if err != nil {
		app.serverError(w, r, err)
		return
	}
	if len(subs) == 0 {
		writeError(w, http.StatusNotFound, "no push subscription registered for your account — enable push notifications in Paramètres first")
		return
	}

	ctx, cancel := context.WithTimeout(r.Context(), pushSendTimeout)
	defer cancel()
	app.sendToUsers(ctx, []int64{user.ID}, pushPayload{
		Title: "Trakka",
		Body:  "Notification de test — si vous voyez ceci, les notifications push fonctionnent.",
		URL:   "/",
	})

	writeJSON(w, http.StatusOK, map[string]any{"sent_to_subscriptions": len(subs)})
}

// ---------------------------------------------------------------------------
// Delivery
// ---------------------------------------------------------------------------

// pushPayload is the JSON body every notification this app sends carries —
// static/sw.js's 'push' listener reads exactly these three fields to call
// self.registration.showNotification. URL is where notificationclick
// navigates to (a deep link of the shape /?list={id} — see
// handleDeepLinkOrRestore in static/js/app.js).
type pushPayload struct {
	Title string `json:"title"`
	Body  string `json:"body"`
	URL   string `json:"url"`
}

// sendToUsers delivers payload to every subscription belonging to any of
// userIDs, concurrently (a household has at most a handful of members, each
// with at most a few devices, so no worker-pool bound is needed) and
// best-effort: one recipient's failed/unreachable push never blocks or
// fails another's, and this function itself never returns an error — a
// push notification is always a secondary effect of some other action that
// has already succeeded by the time this is called (see notifyListChange
// and RunRecurringDueScan below), and must never be able to make that
// action look like it failed. A subscription the push service reports as
// permanently gone (webpush.ErrSubscriptionGone — 404/410) is deleted so
// future notifications stop trying it; any other failure is just logged.
func (app *Application) sendToUsers(ctx context.Context, userIDs []int64, payload pushPayload) {
	if !app.Config.PushEnabled() || len(userIDs) == 0 {
		return
	}
	body, err := json.Marshal(payload)
	if err != nil {
		app.Logger.Error("marshaling push payload", "error", err)
		return
	}

	subs, err := app.DB.ListPushSubscriptionsForUsers(ctx, userIDs)
	if err != nil {
		app.Logger.Error("listing push subscriptions", "error", err)
		return
	}

	keys := webpush.VAPIDKeyPair{PublicKeyB64: app.Config.VAPIDPublicKey, PrivateKeyB64: app.Config.VAPIDPrivateKey}

	var wg sync.WaitGroup
	for _, sub := range subs {
		wg.Add(1)
		go func(sub *models.PushSubscription) { // #nosec G118 -- the context.Background() inside is deliberately given its own subscriptionCleanupTimeout rather than reusing ctx, which may already be at or past its own pushSendTimeout deadline by the time this cleanup runs; see the subscriptionCleanupTimeout doc comment
			defer wg.Done()
			err := webpush.Send(ctx, webpush.Subscription{Endpoint: sub.Endpoint, P256dh: sub.P256dh, Auth: sub.Auth}, keys, app.Config.VAPIDSubject, body)
			if errors.Is(err, webpush.ErrSubscriptionGone) {
				delCtx, delCancel := context.WithTimeout(context.Background(), subscriptionCleanupTimeout)
				defer delCancel()
				if delErr := app.DB.DeletePushSubscriptionByID(delCtx, sub.ID); delErr != nil {
					app.Logger.Error("deleting gone push subscription", "subscription_id", sub.ID, "error", delErr)
				}
				return
			}
			if err != nil {
				app.Logger.Debug("sending push notification failed", "subscription_id", sub.ID, "error", err)
			}
		}(sub)
	}
	wg.Wait()
}

// notifyListChange fires a push notification to every other user with
// access to list (see db.ListNotificationRecipients — House members, plus
// anyone the list or its parent Space has been shared with) when actor adds
// or checks off an item, per this feature's "Use Case 1". Called from
// handleItemsCreate/handleItemsUpdate/handleItemsPatch (items.go) in a
// detached goroutine on its own bounded context — never r.Context(), which
// is canceled the moment the response is written, the same reasoning
// scrapeProductInfo (scrape.go) already follows for its own background
// work — so a slow push delivery can never delay the response to the
// request that triggered it.
//
// The notification text is always composed in French, regardless of the
// recipient's own language preference: although users.language (see
// models.User.Language) now exists, wiring push notification text through
// per-recipient localization is out of scope here — the same reasoning that
// already leaves templates/login.html French-only, per CLAUDE.md's "UI
// language" convention.
func (app *Application) notifyListChange(list *models.List, actor *models.User, itemTitle string, checkedOff bool) {
	go func() {
		ctx, cancel := context.WithTimeout(context.Background(), pushSendTimeout)
		defer cancel()

		recipients, err := app.DB.ListNotificationRecipients(ctx, list.ID, actor.ID)
		if err != nil {
			app.Logger.Error("listing notification recipients", "list_id", list.ID, "error", err)
			return
		}
		if len(recipients) == 0 {
			return
		}

		actorName := actor.DisplayName
		if actorName == "" {
			actorName = actor.Email
		}
		verb := "a ajouté un article"
		if checkedOff {
			verb = "a coché un article"
		}
		payload := pushPayload{
			Title: list.Name,
			Body:  fmt.Sprintf("%s %s (« %s ») dans %s", actorName, verb, itemTitle, list.Name),
			URL:   fmt.Sprintf("/?list=%d", list.ID),
		}
		app.sendToUsers(ctx, recipients, payload)
	}()
}

// ---------------------------------------------------------------------------
// Recurring task due-date reminders ("Use Case 2")
// ---------------------------------------------------------------------------

// recurringDueScanTimeout bounds one whole periodic scan across every
// eligible item — generous for the modest item counts this app targets
// (CLAUDE.md's <20MB RAM footprint implies a small-household scale
// throughout), while still guaranteeing the scan can't hang the process
// indefinitely if something goes wrong partway through.
const recurringDueScanTimeout = 5 * time.Minute

// RunRecurringDueScan checks every recurring, not-done item with a due date
// (see db.ListItemsForRecurringNotifyScan) and sends a reminder push to
// every user with access to its list once the current time is within its
// lead time of that due date — NOTIF_RECURRING_TASK_LEAD_TIME
// (internal/config) by default, or the item's own recurrence_lead_minutes
// override if it has one. Called on a timer from cmd/server/main.go; also
// safe to call directly for an immediate, whole-catalog scan. Best-effort
// throughout, mirroring RunPriceAlertScan: one item's failure (a bad due
// date, a db error resolving its list/recipients) is logged and never stops
// the rest of the scan.
func (app *Application) RunRecurringDueScan(ctx context.Context) {
	if !app.Config.PushEnabled() {
		return
	}

	scanCtx, cancel := context.WithTimeout(ctx, recurringDueScanTimeout)
	defer cancel()

	candidates, err := app.DB.ListItemsForRecurringNotifyScan(scanCtx)
	if err != nil {
		app.Logger.Error("listing items for recurring due scan", "error", err)
		return
	}

	app.Logger.Info("running recurring due-date notification scan", "item_count", len(candidates))
	now := time.Now().UTC()
	for _, c := range candidates {
		if scanCtx.Err() != nil {
			return
		}
		if err := app.checkItemForRecurringDue(scanCtx, c, now); err != nil {
			app.Logger.Error("recurring due scan check failed", "item_id", c.ItemID, "error", err)
		}
	}
}

// checkItemForRecurringDue evaluates one candidate: if now is already
// within its (effective) lead time of its due date, it notifies every user
// with access to its list and records that the reminder was sent for this
// exact due date (db.MarkRecurringReminderSent) so the next scan tick
// doesn't repeat it — see that method's own comment for why storing the due
// date value itself, rather than a plain boolean, is what makes this
// automatically re-arm once the item's due date next changes.
func (app *Application) checkItemForRecurringDue(ctx context.Context, c *db.RecurringDueCandidate, now time.Time) error {
	dueDate, err := time.Parse("2006-01-02", c.DueDate)
	if err != nil {
		return fmt.Errorf("parsing due date %q: %w", c.DueDate, err)
	}

	leadTime := app.Config.NotifRecurringLeadTime
	if c.LeadMinutes != nil {
		leadTime = time.Duration(*c.LeadMinutes) * time.Minute
	}
	if now.Before(dueDate.Add(-leadTime)) {
		return nil // not due soon enough yet
	}

	list, err := app.DB.GetList(ctx, c.ListID)
	if errors.Is(err, db.ErrNotFound) {
		// The list was deleted after this item was fetched by the scan —
		// nothing to notify about, and ON DELETE CASCADE means the item
		// itself is already gone too.
		return nil
	}
	if err != nil {
		return fmt.Errorf("loading list %d: %w", c.ListID, err)
	}

	// Every House member, plus every share recipient, is notified here —
	// unlike notifyListChange there is no single "actor" to exclude, since a
	// due-date reminder is not the result of anyone's own action.
	recipients, err := app.DB.ListNotificationRecipients(ctx, list.ID, 0)
	if err != nil {
		return fmt.Errorf("listing notification recipients: %w", err)
	}
	// A house has exactly one owner at minimum, so a list with zero
	// recipients here would mean list.HouseID itself has no members left —
	// not expected in practice, but if it ever happens there is nobody to
	// notify and nothing further to do.
	if len(recipients) > 0 {
		app.sendToUsers(ctx, recipients, pushPayload{
			Title: "Rappel de tâche",
			Body:  fmt.Sprintf("« %s » (%s) arrive à échéance le %s", c.Title, list.Name, c.DueDate),
			URL:   fmt.Sprintf("/?list=%d", list.ID),
		})
	}

	if err := app.DB.MarkRecurringReminderSent(ctx, c.ItemID, c.DueDate); err != nil {
		return fmt.Errorf("marking reminder sent: %w", err)
	}
	return nil
}
